sdks/python: enable recursive deletion for GCSFileSystem Paths

mohamedawnallah opened the pull request:
## Description

In this PR, we enable recursive deletion for Google Cloud Storage (GCS) paths, including directories and blobs. The `delete` method is updated to remove all blobs under a directory (prefix) when deleting GCS directories. 

Additionally, we update the delete test to verify the recursive deletion of directories containing multiple files.

Changes include:
- Enhanced `delete` method to support recursive directory deletion.
- Added tests for recursive deletion of directories and files.

Fixes #27605

## Current Behavior

Before             |  After
:-------------------------:|:-------------------------:
![before](https://github.com/user-attachments/assets/ac6b70f9-39e9-428d-ab24-ab676bbdb427) |  ![after](https://github.com/user-attachments/assets/d8d2a840-285b-42f2-9498-f0c4e1bee02c)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/actions/workflows/build_wheels.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/actions/workflows/python_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/actions/workflows/java_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/actions/workflows/go_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

mohamedawnallah commented:
Hi @liferoad, It seems the test cases passed now. I would love to receive any feedback on that PR! 🙏

mohamedawnallah commented:
Hi @shunping,  I understand you've been busy last week. Any updates on this PR review? 🙏

shunping commented on sdks/python/apache_beam/io/gcp/gcsio.py:
```diff
@@ -247,13 +247,35 @@ def open(
+
+    # Check if the blob is a directory (prefix) by listing objects
+    # under that prefix.
+    blobs = list(bucket.list_blobs(prefix=blob_name))
```
I am afraid this line could potentially impact the performance of existing pipelines. Particularly, we now have an extra HTTP request to GCS to list blobs in a bucket no matter what. If an existing pipeline is using gcsio.delete() to delete a directory with a large number of files, then it will double the HTTP requests to GCS, which may lead to a request quota exceeded error. 

I think a safer approach is to add a new api in gcsio particularly for this function, then change delete() in gcsfilesystem.py to use this api. Notice that the original issue #27605 is related to the behavior of delete() under gcsfilesystem.py.

mohamedawnallah replied:
> I think a safer approach is to add a new api in gcsio particularly for this function, then change delete() in gcsfilesystem.py to use this api.

Makes sense 👍

> Notice that the original issue https://github.com/apache/beam/issues/27605 is related to the behavior of delete() under gcsfilesystem.py. We can then recommend users to use this api while deleting a directory.

Ooops missed that. I will try to use the safer approach you suggested and ask for review soon :)

mohamedawnallah replied:
> If an existing pipeline is using gcsio.delete() to delete a directory with a large number of files, then the change will double the HTTP requests to GCS, which may lead to a request quota exceeded error.

Makes sense 👍

> Notice that the original issue https://github.com/apache/beam/issues/27605 is related to the behavior of delete() under gcsfilesystem.py.

Ooops missed that. 🙏

> I think a safer approach is to add a new api in gcsio particularly for this function, then change delete() in gcsfilesystem.py to use this api.  We can then recommend users to use this api while deleting a directory.

How about adding an optional parameter, such as `recursive`, with a default value of `false` to ensure backward compatibility and avoid any performance impact on existing pipelines using the `delete` function in `gcsfilesystem.py`?

mohamedawnallah replied:
> How about adding an optional parameter, such as recursive, with a default value of false to ensure backward compatibility and avoid any performance impact on existing pipelines using the delete function in gcsfilesystem.py?

@shunping What do you think about this comment?

shunping replied:
Adding the recursive parameter sounds good to me as it extends the functionality of `gcsio.delete()`.

To fix https://github.com/apache/beam/issues/27605, however, you will also need to make changes to `gcsfilesystem.py` to leverage the new functionality.

For example, you can do something similar to https://github.com/apache/beam/pull/29477/files in `gcsfilesystem.delete()`

```python
  for path in paths:
      if path.endswith('/'):
        # This is a directory. Remove all of its contents including
        # objects and subdirectories.
        self._gcsIO().delete(path, recursive=True)
        return
      else:
         ...
```

shunping commented:
> Hi @shunping, I understand you've been busy last week. Any updates on this PR review? 🙏

Sorry for the late reply, @mohamedawnallah. I left a comment about potential performance impact with the new change.
Since gcsio has been used heavily in Beam, we have to be very careful about any code change to its public api.

shunping reviewed (changes requested):
Please make the changes and I am happy to review them again when ready.

shunping reviewed (changes requested):
Thanks. I think some changes are still needed in gcsfilesystem.py.

shunping commented on sdks/python/apache_beam/io/gcp/gcsfilesystem.py:
```diff
@@ -354,7 +354,7 @@ def delete(self, paths):
     for path in paths:
       if path.endswith('/'):
-        path_to_use = path + '*'
+        return self._gcsIO().delete(path, recursive=True)
```
Actually, that was my mistake when i suggested the code change at https://github.com/apache/beam/pull/33611#discussion_r1937608203.

I think we will need to continue the loop in case users are calling this api to delete multiple directories.

shunping reviewed (approved):
LGTM! Thanks again!

mohamedawnallah commented:
Hi @shunping,  

Thanks for your follow-up review! I've addressed your feedback.

This might be unrelated, but I’d like to use this PR as an opportunity to learn more about Beam development practices as I am getting used to it. My interest was sparked by this discussion: https://github.com/apache/beam/pull/33672#issuecomment-2630967193.

When addressing a review and make a follow-up commit to fix the issues, Someone should avoid squashing reviewed and unreviewed commits. After the follow-up review is completed, would it be necessary to squash the follow-up commit (I am thinking about delay in the merging process e.g for the CI to be triggered again), or is it acceptable to leave the follow-up commit as part of the commit history?

shunping commented:
> Hi @shunping,
> 
> Thanks for your follow-up review! I've addressed your feedback.
> 
> This might be unrelated, but I’d like to use this PR as an opportunity to learn more about Beam development practices as I am getting used to it. My interest was sparked by this discussion: [#33672 (comment)](https://github.com/apache/beam/pull/33672#issuecomment-2630967193).
> 
> When addressing a review and make a follow-up commit to fix the issue, Someone should avoid squashing reviewed and unreviewed commits. After the follow-up review is completed, would it be necessary to squash the follow-up commit (I am thinking about delay in the merging process e.g for the CI to be triggered again), or is it acceptable to leave the follow-up commit as part of the commit history?

The root problem here is about modifying an "already pushed" commit. In your previous PR and this one as well, you pushed your code to github and then you did an operation locally to modify your commit history (eg. squashing the commits or simply amending an already pushed commit). You may notice that, when you pushed your code again, git reminded you to do a "forced" push, which is then shown up in github as something like:
```
mohamedawnallah force-pushed the enableRecursiveDeleteGCS branch from 661f822 to 3491c70
```

This is not a good practice for code that is already in the reviewing pipeline. Specifically, when you force-push a commit (which changes the original commit sha1), github can no longer associate it with the reviews that was previously submitted to it. As a result, it will be inconvenient for reviewers to go back and forth to compare what you had changed and what were the previous feedback.

Check the following pages for some more reading.
- https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#about-pull-requests
- https://jvns.ca/blog/2023/11/06/rebasing-what-can-go-wrong-/#force-pushing-makes-code-reviews-harder

liferoad commented:
Can we add this to CHANGES.md?

shunping commented:
Thank you!

liferoad commented on CHANGES.md:
```diff
@@ -71,6 +71,7 @@
 * Add BigQuery vector/embedding ingestion and enrichment components to apache_beam.ml.rag (Python) ([#33413](https://github.com/apache/beam/pull/33413)).
 * Upgraded to protobuf 4 (Java) ([#33192](https://github.com/apache/beam/issues/33192)).
 * [GCSIO] Added retry logic to each batch method of the GCS IO (Python) ([#33539](https://github.com/apache/beam/pull/33539))
+* [GCSIO] enable recursive deletion for GCSFileSystem Paths (Python) ([#33611](https://github.com/apache/beam/pull/33611)).
```
Enable?
